### PR TITLE
Add wheel builds to cuxfilter

### LINKED ISSUE #491: [IMPROVEMENT] Add wheel builds to cuxfilter
Follow [this](https://github.com/rapidsai/ops/issues/2572) guide to add wheels builds to cuxfilter.

- [x] Transition CI to GHA from Jenkins: The wheel builds use Github Actions.
- [x] Remove usage of versioneer: Various aspects of versioneer's operation and generated versions are incompatible with other parts of the Python ecosystem.
- [x] Migrate as much logic as possible from setup.py to pyproject.toml
- [ ] Add wheels GHA workflows

cc @exactlyallan @bdice 

edit: updated checklist as per suggestion

### PR DESCRIPTION
**Pull request by AjayThorve**
This PR aims to  add wheels builds to cuxfilter

- [x] remove versioneer support
- [x] move build logic to pyproject.toml
- [x] add GHA to build, test and publish wheels
- [x] Update dependencies.yaml

Resolves #491
cc @bdice @exactlyallan

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -2,17 +2,25 @@
...
         hooks:
               - id: flake8
-                args: ["--config=python/setup.cfg"]
+                args: ["--config=.flake8"]
                 files: python/.*$
+                types: [file]
+                types_or: [python, cython]
+                additional_dependencies: ["flake8-force"]
+      - repo: https://github.com/rapidsai/dependency-file-generator
+        rev: v1.5.1
```
Comment (ajschmidt8) [inline .pre-commit-config.yaml:21]:
```suggestion
        rev: v1.5.2
```

`1.5.2` is available: https://github.com/rapidsai/dependency-file-generator/releases/tag/v1.5.2

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -57,3 +57,23 @@ jobs:
...
       sha: ${{ inputs.sha }}
+  wheel-build:
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/wheels-pure-build.yml@cuda-120
+    with:
+      build_type: ${{ inputs.build_type || 'branch' }}
+      branch: ${{ inputs.branch }}
+      sha: ${{ inputs.sha }}
+      date: ${{ inputs.date }}
+      package-name: cudf
```
Comment (bdice) [inline .github/workflows/build.yaml:68]:
```suggestion
      package-name: cuxfilter
```

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -57,3 +57,23 @@ jobs:
...
+      package-dir: python/cuxfilter
```
Comment (bdice) [inline .github/workflows/build.yaml:69]: Try this, it might help. The `python/` directory contains the `pyproject.toml`. Note that I saw issues locally if I didn't have the trailing slash.
```suggestion
      package-dir: python/
```

File: .github/workflows/pr.yaml [hunk 4]
```diff
@@ -55,3 +57,19 @@ jobs:
...
       container_image: "rapidsai/ci:latest"
       run_script: "ci/build_docs.sh"
+  wheel-build:
+    needs: checks
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/wheels-pure-build.yml@cuda-120
+    with:
+      build_type: pull-request
+      package-name: cuxfilter
+      package-dir: python
```
Comment (bdice) [inline .github/workflows/pr.yaml:67]: Let's fix this one too, since it's the one being tested in PRs. :)
```suggestion
      package-dir: python/
```

File: .flake8 [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+[flake8]
+filename = *.py,
+exclude = __init__.py, *.egg, build, docs, .git
+force-check = True
```
Comment (bdice) [inline .flake8:6]: The `force-check` option is only needed for Cython libraries that want to enable flake8 support. We can drop it here since there's no Cython. See: https://github.com/kmaehashi/flake8-force

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -2,17 +2,25 @@
 
 repos:
       - repo: https://github.com/psf/black
-        rev: 22.10.0
+        rev: 22.3.0
```
Comment (bdice) [inline .pre-commit-config.yaml:5]: Did we downgrade this for a reason?

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -2,17 +2,25 @@
...
         hooks:
               - id: black
                 files: python/cuxfilter/.*
                 # Explicitly specify the pyproject.toml at the repo root, not per-project.
                 args: ["--config", "pyproject.toml"]
       - repo: https://github.com/PyCQA/flake8
-        rev: 6.0.0
+        rev: 5.0.4
```
Comment (bdice) [inline .pre-commit-config.yaml:12]: Did we downgrade this for a reason?

File: .pre-commit-config.yaml [hunk 8]
```diff
@@ -2,17 +2,25 @@
...
         hooks:
               - id: flake8
-                args: ["--config=python/setup.cfg"]
+                args: ["--config=.flake8"]
                 files: python/.*$
+                types: [file]
+                types_or: [python, cython]
+                additional_dependencies: ["flake8-force"]
```
Comment (bdice) [inline .pre-commit-config.yaml:19]: All this can be removed since we don't need Cython support.
```suggestion
```

File: ci/release/apply_wheel_modifications.sh [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+# cuxfilter pyproject.toml cuda suffixes
+sed -i "s/^name = \"cuxfilter\"/name = \"cuxfilter${CUDA_SUFFIX}\"/g" python/pyproject.toml
+# Need to provide the == to avoid modifying the URL
+sed -i "s/\"cudf==/\"cudf${CUDA_SUFFIX}==/g" python/pyproject.toml
+sed -i "s/\"cuspatial==/\"cuspatial${CUDA_SUFFIX}==/g" python/pyproject.toml
+sed -i "s/\"dask-cuda==/\"dask-cuda${CUDA_SUFFIX}==/g" python/pyproject.toml
+sed -i "s/\"dask-cudf==/\"dask-cudf${CUDA_SUFFIX}==/g" python/pyproject.toml
+
+if [[ $CUDA_SUFFIX == "-cu12" ]]; then
+    sed -i "s/numba[<=>\.,0-9]*/numba>=0.57/g" python/pyproject.toml
```
Comment (bdice) [inline ci/release/apply_wheel_modifications.sh:21]: This shouldn't be needed -- cuxfilter requires `numba >=0.57` as of #480.
```suggestion
```

File: conda/recipes/cuxfilter/meta.yaml [hunk 10]
```diff
@@ -25,11 +25,11 @@ requirements:
     - cuda-version ={{ cuda_version }}
   run:
     - bokeh >=2.4.2,<=2.5
-    - cudf ={{ minor_version }}
+    - cudf ={{ version }}
```
Comment (bdice) [inline conda/recipes/cuxfilter/meta.yaml:28]: Why was this changed? I think the previous form was correct. We don't want to pin on the full version because patches aren't necessarily released for every RAPIDS package.
```suggestion
    - cudf ={{ minor_version }}
```

File: conda/recipes/cuxfilter/meta.yaml [hunk 11]
```diff
@@ -25,11 +25,11 @@ requirements:
...
     - cupy >=12.0.0
-    - cuspatial ={{ minor_version }}
-    - dask-cuda ={{ minor_version }}
-    - dask-cudf ={{ minor_version }}
+    - cuspatial ={{ version }}
+    - dask-cuda ={{ version }}
+    - dask-cudf ={{ version }}
```
Comment (bdice) [inline conda/recipes/cuxfilter/meta.yaml:32]:
```suggestion
    - cuspatial ={{ minor_version }}
    - dask-cuda ={{ minor_version }}
    - dask-cudf ={{ minor_version }}
```

File: dependencies.yaml [hunk 12]
```diff
@@ -140,10 +169,13 @@ dependencies:
           - cupy>=12.0.0
       - output_types: requirements
```
Comment (bdice) [inline dependencies.yaml:170]: Due to `--extra-index-url` reasons for ptxcompiler/cubinlinker, cudf wasn't able to combine these. Luckily, cuxfilter can combine them.
```suggestion
      - output_types: [requirements, pyproject]
```

File: dependencies.yaml [hunk 13]
```diff
@@ -140,10 +169,13 @@ dependencies:
...
         packages:
-          - cupy-cuda11x>=12.0.0
+          - &cupy_pip cupy-cuda11x>=12.0.0
+      - output_types: pyproject
+        packages:
+          - *cupy_pip
```
Comment (bdice) [inline dependencies.yaml:175]: See previous suggestion.
```suggestion
          - cupy-cuda11x>=12.0.0
```

File: python/pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    "cupy-cuda11x>=12.0.0",
+    "cuspatial==23.8.*",
+    "dask-cuda==23.8.*",
+    "dask-cudf==23.8.*",
+    "datashader>=0.15",
+    "geopandas>=0.11.0",
+    "holoviews>=1.15.0,<=1.15.4",
+    "jupyter-server-proxy",
+    "libwebp",
+    "nodejs>=14",
```
Comment (bdice) [inline python/pyproject.toml:32]: This seems incorrect. The PyPI package for `nodejs` isn't what you want here. https://pypi.org/project/nodejs/

I think `nodejs` is supposed to be a conda-only dependency.

File: python/setup.py [hunk 15]
```diff
@@ -1,36 +1,10 @@
 from setuptools import find_packages, setup
```
Comment (bdice) [inline python/setup.py:1]:
```suggestion
# Copyright (c) 2019-2023, NVIDIA CORPORATION.

from setuptools import find_packages, setup
```

File: python/cuxfilter/__init__.py [hunk 16]
```diff
@@ -1,6 +1,4 @@
 from cuxfilter.dataframe import DataFrame
```
Comment (bdice) [inline python/cuxfilter/__init__.py:1]:
```suggestion
# Copyright (c) 2019-2023, NVIDIA CORPORATION.

from cuxfilter.dataframe import DataFrame
```

**Review by bdice — COMMENTED** [review]
Suggestions attached -- nice work @AjayThorve, really glad to have these updates.

File: .flake8 (hunk 5 above)
Comment (bdice) [reply .flake8:6]:
```suggestion
```

File: python/pyproject.toml (hunk 14 above)
Comment (bdice) [reply python/pyproject.toml:32]: Probably the same issue exists for `libwebp` (no PyPI package, probably conda-only).

File: .pre-commit-config.yaml (hunk 6 above)
Comment (AjayThorve) [reply .pre-commit-config.yaml:5]: @bdice For all of the versions in this file, I was following what cuDF had. It's actually different for cuML and cuDF. Is there a fixed standard of versions we are following?

File: python/pyproject.toml (hunk 14 above)
Comment (AjayThorve) [reply python/pyproject.toml]: that's a good point. I have moved it to conda specific requirements

File: dependencies.yaml (hunk 12 above)
Comment (AjayThorve) [reply dependencies.yaml]: fixed it

File: dependencies.yaml (hunk 13 above)
Comment (AjayThorve) [reply dependencies.yaml]: fixed

File: conda/recipes/cuxfilter/meta.yaml (hunk 10 above)
Comment (AjayThorve) [reply conda/recipes/cuxfilter/meta.yaml]: reverted the change

File: conda/recipes/cuxfilter/meta.yaml (hunk 11 above)
Comment (AjayThorve) [reply conda/recipes/cuxfilter/meta.yaml]: reverted the change

File: .pre-commit-config.yaml (hunk 6 above)
Comment (bdice) [reply .pre-commit-config.yaml:5]: Since transitioning to GitHub Actions, we have decoupled these versions from the `integration` repo. As a result, every RAPIDS repo is free to upgrade linter versions at any time. You can update all the linters automatically with `pre-commit autoupdate` or pick your favorite upstream repo (e.g. cudf) and align version numbers, up to you. It's been my intention to run `pre-commit autoupdate` across all of RAPIDS for ... at least 6 months now, but that's been a low priority.

Comment (bdice) [reply .pre-commit-config.yaml:5]: Note that there is already some heterogeneity, such as cuml doesn't run `isort` due to having some circular dependency / import ordering issues. I would advocate for repo maintainers to run `pre-commit autoupdate` on a regular basis, maybe every release or two.

Comment (AjayThorve) [reply .pre-commit-config.yaml]: Thanks @bdice, updated it using `pre-commit autoupdate`. Will keep in mind to update every once in a while.

File: .pre-commit-config.yaml (hunk 7 above)
Comment (AjayThorve) [reply .pre-commit-config.yaml]: fixed

**Review by bdice — APPROVED** [review]
LGTM!

File: .github/workflows/pr.yaml [hunk 17]
```diff
@@ -55,3 +57,19 @@ jobs:
...
+      package-name: cuxfilter
+      package-dir: python/
+  wheel-tests:
+    needs: wheel-build
+    secrets: inherit
+    uses: rapidsai/shared-action-workflows/.github/workflows/wheels-pure-test.yml@cuda-120
+    with:
+      build_type: pull-request
+      package-name: cuxfilter
+      test-unittest: "python -m pytest -n 8 ./python/cuxfilter/tests"
```
Comment (bdice) [inline .github/workflows/pr.yaml:75]: You'll need to borrow some logic from here for dask-cuda: https://github.com/rapidsai/cudf/blob/1899fb1990eae693bd74f92fa4a3f6a217889210/.github/workflows/pr.yaml#L132-L133

dask-cuda has to install from source, since the `23.8.*a*` packages on PyPI are pre-release / alphas. We can't really use `pip install ... --pre` because it uses pre-release versions of all packages.

I'm not sure if you need _just_ dask-cuda or also dask/distributed.

Also note that you'll need to change `update-version.sh` to accomodate this. https://github.com/rapidsai/cudf/blob/1899fb1990eae693bd74f92fa4a3f6a217889210/ci/release/update-version.sh#L112

All of dask/distributed/dask-cuda:
```suggestion
      # Test against latest dask/distributed/dask-cuda.
      test-before: "python -m pip install git+https://github.com/dask/dask.git@main git+https://github.com/dask/distributed.git@main git+https://github.com/rapidsai/dask-cuda.git@branch-23.08"
      test-unittest: "python -m pytest -n 8 ./python/cuxfilter/tests"
```

Only dask-cuda:
```suggestion
      # Test against latest dask-cuda.
      test-before: "python -m pip install git+https://github.com/rapidsai/dask-cuda.git@branch-23.08"
      test-unittest: "python -m pytest -n 8 ./python/cuxfilter/tests"
```

File: ci/release/update-version.sh
Comment (bdice) [inline ci/release/update-version.sh:1]: Put
```
sed_runner "s/dask-cuda.git@branch-[^\"\s]\+/dask-cuda.git@branch-${NEXT_SHORT_TAG}/g" ${FILE};
```
just below line 63. See https://github.com/rapidsai/cuxfilter/pull/497/files#r1260365685.

File: .github/workflows/pr.yaml (hunk 17 above)
Comment (AjayThorve) [reply .github/workflows/pr.yaml:75]: ah okay got it. It would just be dask-cuda at this point. Thanks again!

@bdice I am assuming this is a temporary fix, until dask-cuda has a pip release (23.10.*?), right?

File: ci/release/update-version.sh
Comment (AjayThorve) [reply ci/release/update-version.sh:1]: done

File: .github/workflows/pr.yaml (hunk 17 above)
Comment (AjayThorve) [reply .github/workflows/pr.yaml:75]: Actually @bdice, we don't need dask_cuda as a direct dependency, since it's only used in the dask_cudf based notebooks. So I removed the direct dependency, and added it to `notebooks` dependency only

Comment (bdice) [reply .github/workflows/pr.yaml:75]: Great that the dependency could be moved.

To answer your question about whether this would be a temporary fix, it isn't temporary. The dask-cuda wheels used in CI are always pre-release alphas. We are always testing RAPIDS packages on `branch-XX.YY` against pre-releases of dask-cuda `XX.YY` to ensure that when the releases are made, everything works together as intended.

**Review by bdice — APPROVED** [review]

File: .github/workflows/pr.yaml (hunk 17 above)
Comment (AjayThorve) [reply .github/workflows/pr.yaml:75]: got it, good to know. Thanks again for the thorough review. Let me know if this looks good to merge.

**Review by raydouglass — APPROVED** [review]

**Review by exactlyallan — APPROVED** [review]

**Comment by AjayThorve** [comment]
Thanks everyone for the reviews!
